### ISSUE TITLE
New Roslyn performance analyzer: avoid captures

### ISSUE THREAD

**Issue by sharwell** [dotnet/roslyn-analyzers#1438]
An analyzer should be added so code marked with `[PerformanceSensitive]` reports a warning if it captures variables. See dotnet/roslyn#23554 for an example of the kind of code this analyzer aim to protect.

**Comment by AkbarDizaji** [comment]
Hi! I'd like to pick this up. Before starting, I did a gap analysis of this issue against the current code (which now living in `dotnet/roslyn` under `src/RoslynAnalyzers/PerformanceSensitiveAnalyzers` after the migration):

- `DisplayClassAllocationAnalyzer` (HAA0301/HAA0302) already covers the core  of this request for lambdas and anonymous method, it registers `ParenthesizedLambdaExpression`, `SimpleLambdaExpression`, and `AnonymousMethodExpression`, and reports captures via `SemanticModel.AnalyzeDataFlow`.
- However, it does not register `LocalFunctionStatement`. A local function that captures variables and is converted to a delegate
  (e.g. `Action a = LocalFunc;`) gets no closure diagnostic:
  `TypeConversionAllocationAnalyzer` (HAA0603) flags the delegate creation itself, but it has no capture analysis, so nothing reports the captured state / closure allocation.
- Local functions that are only invoked directly correctly produce no diagnostic today (no heap allocation there), and that behavior should be  preserved.
- A few related scenarios also have no test coverage: local functions  (either way), static lambdas, `this`-only capture, expression trees,
  query syntax, nested lambdas.

I'd like to contribute focused coverage for the local-function-to-delegate case, starting with characterization tests for the current behavior.
Three questions first:

1. Is extending the existing HAA0301/HAA0302 rules the preferred direction,  or would you rather see a new rule ID for the local-function case?
2. Since active development moved to `dotnet/roslyn`, should I open a follow-up issue there and send the PR against `dotnet/roslyn`?
3. @genlu this was assigned to you back in 2019 with no linked PR since is work still planned on your side, or is it OK for me to take it?
